cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+performing a simple type check on the entries in the dict, whether the objects
+are instance of `FileFactory` or `Directory`.
+
+Following is the prototype for the `Directory` class.
+
+```python
+import os
+
+from coalib.parsing.Globbing import glob
+
+
+class Directory:
+    def __init__(self, path):
+        self.path = path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self.files = 0 # number of files contained in a directory
+        self.subdirs = 0 # number of sub-directories contained in a directory
+
+    def count_children(self):
+        children = os.listdir(self.path)
+
+        for child in children:
+            if os.path.isfile(child):
+                self.files += 1
+
+        self.subdirs = len(children) - self.files
+
+    def get_dir(pattern):
+        return glob(pattern) # get sub directory matching the given glob pattern
+```
+
+### Ignore directories functionality
+
+This will probably be the biggest performance boost for coala. Instead of
+ignoring only unmodified files on subsequent runs, coala shall also ignore
+directories with no unchanged files and subdirectories. This will free up the
+unnecessary space in the cache and speed up runs. It can be implemented
+accordingly:
+
+- `FileFactory` will have an additional attribute parent
+  (the parent directory path) in its `__init__` method.
+- `Directory` class will also have additional attributes `files`
+  (number of files) `subdirs` (number of subdirs) and `parent`
+  (the parent directory's path; in case of project's root it will be `None`)
+- There will be a separate dict (called directory-tracker) containing directory
+  paths as keys and number of files and subdirs as a list of values for e.g.
+  ```
+      { Directory.path: [Directory.files, Directory.subdirs] }
+  ```
+- First all the unchanged files will be deleted from the cache whenever a file
+  is deleted from the cache, we will look up its parent directory using
+  `FileFactory.parent` and decrease the count of the number of files
+  (`Directory.files` in the directory-tracker dict) corresponding to that path
+  by one.
+- Once all the files have been deleted from the cache coala will move on to the
+  directories in our dict. coala will look for directories with
+  `Directory.subdirs` = 0 (starting from the bottommost level of the directory
+  tree) if the `Directory.files` associated with such paths is 0 (i.e. all the
+  files contained inside that directory have been uncached) then the
+  `Directory` object corresponding to this path will be deleted from the cache.
+- As soon as a Directory object is uncached the value of the
+  `Directory.subdirs` of their parent directory will also be decreased by 1.
+- This process will continue for all the directories and the paths having both
+  the values (`Directory.files` and `Directory.subdirs`) as zero will have
+  their corresponding `Directory` objects deleted from the cache.
```

Comment:
Still a bit vague, but I know this is currently one of our harder parts in this cEP, so we can initially merge it and improve it later on.